src/wallet/wallet.cpp

```diff
@@ -6750,6 +6757,117 @@ bool CWallet::InitLoadWallet(const CChainParams& params, bool clearWitnessCaches
         walletInstance->SetMaxVersion(nMaxVersion);
     }
 
+    auto recoverPath = GetOptionalArg("-recoverwallet");
+    if (recoverPath.has_value()) {
+        if (!fFirstRun) {
+            return UIError(_("Cannot recover over an existing wallet file."));
+        } else if (fs::exists(recoverPath.value())) {
+            // Parse file and search for "recovery_phrase="
+            std::ifstream recoverFile;
+            recoverFile.open(recoverPath.value(), std::ios::in | std::ios::ate);
+            assert(recoverFile.is_open());
+            recoverFile.seekg(0, recoverFile.beg);
+            SecureString line;
+            std::optional<SecureString> recoveryPhrase;
+            std::optional<Language> language;
+            std::optional<uint256> fingerprint;
+            uint32_t numAccounts = 1;
+
+            uint32_t lineNumber = 0;
+            uint32_t lastUsedLine;
+            while (recoverFile.good()) {
+                // Constrain parser to follow the multi-line format of z_exportwallet:
+                std::getline(recoverFile, line);
+                lineNumber++;
+
+                if (!line.empty()) {
+                    auto recoverPos = line.find("recovery_phrase=");
+                    if (recoverPos != std::string::npos) {
+                        // Check that file contains exactly one recovery phrase line
+                        assert(!recoveryPhrase.has_value());
+                        lastUsedLine = lineNumber;
+                        SecureString tmp = line.substr(recoverPos + 16, std::string::npos);
+                        boost::trim(tmp);
+                        recoveryPhrase = boost::algorithm::erase_all_copy(tmp, "\"");
+                    }
+
+                    // If a "language=" line appears, it must immediately follow
+                    // "recovery_phrase="
+                    auto languagePos = line.find("language=");
+                    if (languagePos != std::string::npos) {
+                        assert(recoveryPhrase.has_value());
+                        assert(lineNumber == (lastUsedLine + 1));
```

Comment:
These shouldn't be assertions; if they are needed then they should be user-friendly errors. Note that `zcashd-wallet-tool` is more relaxed in its parsing.